cdap-watchdog/src/main/java/co/cask/cdap/logging/logbuffer/recover/LogBufferRecoveryService.java

```diff
@@ -0,0 +1,156 @@
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Log buffer recovery service which recovers logs upon log saver restart and sends them to log buffer pipeline for
+ * further processing.
+ */
+public class LogBufferRecoveryService extends AbstractExecutionThreadService {
+  private static final Logger LOG = LoggerFactory.getLogger(LogBufferRecoveryService.class);
+  // For outage, only log once per 60 seconds per message.
+  private static final Logger OUTAGE_LOG =
+    Loggers.sampling(LOG, LogSamplers.perMessage(() -> LogSamplers.limitRate(60000)));
+  private static final String SERVICE_NAME = "log.buffer.recovery";
+
+  private final List<LogBufferProcessorPipeline> pipelines;
+  private final List<CheckpointManager<LogBufferFileOffset>> checkpointManagers;
+  private final String baseLogDir;
+  private final int batchSize;
+  private final CountDownLatch stopLatch;
+
+  private LogBufferReader reader;
+  private volatile boolean stopped;
+
+  public LogBufferRecoveryService(CConfiguration cConf, List<LogBufferProcessorPipeline> pipelines,
+                                  List<CheckpointManager<LogBufferFileOffset>> checkpointManagers) {
+    this(pipelines, checkpointManagers, cConf.get(Constants.LogBuffer.LOG_BUFFER_BASE_DIR),
+         cConf.getInt(Constants.LogBuffer.LOG_BUFFER_RECOVERY_BATCH_SIZE));
+  }
+
+  @VisibleForTesting
+  LogBufferRecoveryService(List<LogBufferProcessorPipeline> pipelines,
+                           List<CheckpointManager<LogBufferFileOffset>> checkpointManager,
+                           String baseLogDir, int batchSize) {
+    this.pipelines = pipelines;
+    this.checkpointManagers = checkpointManager;
+    this.baseLogDir = baseLogDir;
+    this.batchSize = batchSize;
+    this.stopLatch = new CountDownLatch(1);
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    // get the smallest offset of all the log pipelines
+    LogBufferFileOffset minOffset = getSmallestOffset(checkpointManagers);
+    this.reader = new LogBufferReader(baseLogDir, batchSize, minOffset.getFileId(), minOffset.getFilePos());
+  }
+
+  @Override
+  protected void run() throws Exception {
+    // check if the log buffer dir exists, this could happen if log saver is starting for the first time.
+    if (!exists(baseLogDir)) {
+      return;
+    }
+
+    List<LogBufferEvent> logBufferEvents = new LinkedList<>();
+    boolean hasReadEvents = true;
+    while (!stopped && hasReadEvents) {
+      try {
+        hasReadEvents = reader.readEvents(logBufferEvents) > 0;
+        recoverLogs(logBufferEvents, pipelines);
+      } catch (IOException e) {
+        // even though error occurred while reading, whatever logs were read, those should be processed.
+        recoverLogs(logBufferEvents, pipelines);
+        OUTAGE_LOG.warn("Failed to recover logs from log buffer. Read will be retried.", e);
+        // in case of failure to read, sleep and then retry
+        stopLatch.await(500, TimeUnit.MILLISECONDS);
```

Comment:
If the call to `recoverLogs` has exception being thrown, then the `logBufferEvents` list will not be getting cleared. If the outage in the logs process pipeline persist, then this can potentially cause out of memory issue.

Reply:
Since recover logs only puts the events in memory queue (similar to concurrent writer), it should not throw any exception. https://github.com/cdapio/cdap/blob/develop/cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java#L178

Reply:
If `recoverLogs` never throw, then it's better to move it outside of the `try` block.

Reply:
This is because when exception occurs `recoverLogs` should be called before sleep.